lib/Zonemaster/Backend/RPCAPI.pm

```diff
@@ -650,6 +637,69 @@ sub get_batch_job_result {
     return $result;
 }
 
+sub _get_locale {
+    my ( $self, $params ) = @_;
+    my @error;
+
+    my $language = $params->{language};
+    my %locales = $self->{config}->LANGUAGE_locale;
+    my $locale;
+
+    if ( !defined $language ) {
+        return $locale, \@error;
+    }
+
+    if ( length $language == 2 ) {
+        if ( !exists $locales{$language} ) {
+            push @error, {
+                path => "/language",
+                message => N__ "Unkown language string"
+            };
+        }
+        elsif ( scalar keys %{ $locales{$language} } > 1 ) {
+            push @error, {
+                path => "/language",
+                message => N__ "Language string not unique"
+            };
+        } else {
+            ( $locale ) = keys %{ $locales{$language} };
+        }
+    }
+    else {
+        if ( !exists $locales{substr $language, 0, 2}{$language} ) {
+            push @error, {
+                path => "/language",
+                message => N__ "Unkown language string"
+            };
+        } else {
+            $locale = $language;
+        }
+    }
+
+    if (defined $locale) {
+        $locale .= '.UTF-8';
+    }
+
+    return $locale, \@error,
+}
+
+sub _set_error_message_locale {
+    my ( $self, $params ) = @_;
+
+    my @error_response = ();
+    my ($locale, $locale_error) = $self->_get_locale( $params );
+    push @error_response, @{$locale_error};
+
+    if (not defined $locale or $locale eq "") {
+        # Don't translate message if locale is not defined
+        $locale = "C";
+    }
+
+    # Use POSIX implementation instead of Locale::Messages wrapper
+    setlocale( LC_ALL, $locale );
+    return @error_response;
+}
```

Comment:
> We've needed all that logic for the platforms we listed as supported. Of course the versions of those OSes have changed over time. I guess maybe the included gettext implementations could have converged in behavior over time, but I wouldn't bet on it.

I understand that, but if it works as is I don't want to add magic instructions that I both don't understand and can't test. That being said I will double check that it works correctly on the listed supported platforms.

> Also this does not follow the calling convention prescribed in the gettext documentation (https://www.gnu.org/software/gettext/manual/html_node/Triggering.html#Triggering). At the very least we should do that.

It should already be the case. The locale is set using `LC_ALL`, `bindtextdomain` is called in `Locale::TextDomain` when the translation happens (`Locale::TextDomain::__` calls `Locale::TextDomain::__find_domain` that calls `bindtextdomain`), finally the call to `textdomain` should be unneeded as `Locale::TextDomain::__` uses `dgettext`.

> Basically what I've done in Engine is I've tried to concentrate all those details and differences behind an interface that works the same on all platforms. I've tried hard to make the interface as simple as possible but I'm not convinced I've found the perfect solution (nor that there is one). This stuff is a lot trickier than it may seem.

I will retest it a bit more thoroughly but it would have helped to know which systems were causing trouble.

Reply:
> I understand that, but if it works as is I don't want to add magic instructions that I both don't understand and can't test. That being said I will double check that it works correctly on the listed supported platforms.

I agree that this is a good stance.

> > Also this does not follow the calling convention prescribed in the gettext documentation (https://www.gnu.org/software/gettext/manual/html_node/Triggering.html#Triggering). At the very least we should do that.
> 
> It should already be the case. The locale is set using `LC_ALL`, `bindtextdomain` is called in `Locale::TextDomain` when the translation happens (`Locale::TextDomain::__` calls `Locale::TextDomain::__find_domain` that calls `bindtextdomain`), finally the call to `textdomain` should be unneeded as `Locale::TextDomain::__` uses `dgettext`.

The difference I see is in the call to `setlocale()`. Not too long ago I've had to deal with some platform(s) supported by Zonemaster that wouldn't accept values given explicitly in the second argument to `setlocale()`. But it worked when I gave an empty string and had `setlocale()` look up the value from the environment. The way to go with setlocale/gettext seems to be to not stray an inch from the documentation if you can avoid it. (Yes, this is terrible API design, but it is what we have to deal with.)

Reply:
From what I see in Engine, we are setting the `LC_MESSAGES` environment variable if the call to `setlocal` wasn't successful, but I don't see where we give it an empty string. Where would that be?

Also Alex and I have noticed that, on linux (tested on Archlinux, my dev environment, and a fresh Debian 11 VM), when the `LANGUAGE` environment is set the translation does not happen, whether it is error message translation or the test results.

I am starting the RPC API with  `LANGUAGE=en_US:en LANG=en_US.UTF-8 ZM_BACKEND_RPCAPI_LOGLEVEL=debug starman --preload-app script/zonemaster_backend_rpcapi.psgi`.

Reply:
I have the same translation issue on test results and error messages on freebsd and rockylinux.

Reply:
I have opened a PR to address the issue ( #894 ), I am not sure if it is the right way to do it but it works fine and fix both translation issues. With that being addressed error message translation seems to be working fine on all supported platforms.

Reply:
I recognize this problem with having LANGUAGE set at backend start from several years back. As I recall it, there was a problem with LANG to. I had a work-around where I unset all locale setting before starting backend.

Reply:
> From what I see in Engine, we are setting the LC_MESSAGES environment variable if the call to setlocal wasn't successful, but I don't see where we give it an empty string. Where would that be?

I was sure the `around locale` modifier in Zonemaster::Engine::Translator did that, but apparently it doesn't now.

Reply:
> I have opened a PR to address the issue ( #894 ), I am not sure if it is the right way to do it but it works fine and fix both translation issues.

AFAICT it is the right way.

Reply:
We used to have the code for setting locale spread out in Engine, CLI and Backend. I worked hard to get the locale setting centralized into Engine. I'd really like to have the error message translation code use the same locale setting code as the test case message translation code. But I guess we can do that later.